service/history/historyEngine.go

```diff
@@ -1081,32 +1085,165 @@ func (e *historyEngineImpl) DescribeMutableState(
 
 	if cacheHit && cacheCtx.(*workflowExecutionContextImpl).mutableState != nil {
 		msb := cacheCtx.(*workflowExecutionContextImpl).mutableState
-		response.CacheMutableState, err = e.toMutableStateJSON(msb)
-		if err != nil {
-			return nil, err
-		}
+		wms := msb.CopyToPersistence()
+		response.CacheMutableState = e.workflowMutableStateToJSON(wms)
 	}
 
 	msb, err := dbCtx.loadWorkflowExecution()
 	if err != nil {
-		return nil, err
+		response.DatabaseMutableState = err.Error()
+		return response, nil
+	}
+	wms := msb.CopyToPersistence()
+	response.DatabaseMutableState = e.workflowMutableStateToJSON(wms)
+
+	currentBranchToken := wms.ExecutionInfo.EventBranchToken
+	if wms.VersionHistories != nil {
+		// if VersionHistories is set, then all branch infos are stored in VersionHistories
+		currentVersionHistory, err := wms.VersionHistories.GetCurrentVersionHistory()
+		if err != nil {
+			response.TreeId = err.Error()
+			return response, nil
+		}
+		currentBranchToken = currentVersionHistory.GetBranchToken()
 	}
-	response.DatabaseMutableState, err = e.toMutableStateJSON(msb)
+	branchInfo := persistenceblobs.HistoryBranch{}
+	err = branchInfo.Unmarshal(currentBranchToken)
 	if err != nil {
-		return nil, err
+		response.TreeId = err.Error()
+		return response, nil
 	}
 
+	response.TreeId = branchInfo.TreeId
+	response.BranchId = branchInfo.BranchId
+
 	return response, nil
 }
 
-func (e *historyEngineImpl) toMutableStateJSON(msb mutableState) (string, error) {
-	ms := msb.CopyToPersistence()
+func (e *historyEngineImpl) workflowMutableStateToJSON(wms *persistence.WorkflowMutableState) string {
```

Comment:
Can you move this into common helpers?  I don't like the imports it brings into this file.